x-pack/plugins/enterprise_search/server/plugin.ts

```diff
@@ -0,0 +1,133 @@
+import { appSearchTelemetryType } from './saved_objects/app_search/telemetry';
+
+export interface PluginsSetup {
+  usageCollection?: UsageCollectionSetup;
+  security?: SecurityPluginSetup;
+  features: FeaturesPluginSetup;
+}
+
+export interface ServerConfigType {
+  host?: string;
+  enabled: boolean;
+  accessCheckTimeout: number;
+  accessCheckTimeoutWarning: number;
+}
+
+export interface IRouteDependencies {
+  router: IRouter;
+  config: ServerConfigType;
+  log: Logger;
+  getSavedObjectsService?(): SavedObjectsServiceStart;
+}
+
+export class EnterpriseSearchPlugin implements Plugin {
+  private config: Observable<ServerConfigType>;
+  private logger: Logger;
+
+  constructor(initializerContext: PluginInitializerContext) {
+    this.config = initializerContext.config.create<ServerConfigType>();
+    this.logger = initializerContext.logger.get();
+  }
+
+  public async setup(
+    { capabilities, http, savedObjects, getStartServices }: CoreSetup,
+    { usageCollection, security, features }: PluginsSetup
+  ) {
+    const config = await this.config.pipe(first()).toPromise();
+
+    /**
+     * Register space/feature control
+     */
+    features.registerFeature({
+      id: 'enterpriseSearch',
+      name: 'Enterprise Search',
+      order: 0,
+      icon: 'logoEnterpriseSearch',
+      navLinkId: 'appSearch', // TODO - remove this once functional tests no longer rely on navLinkId
+      app: ['kibana', 'appSearch'], // TODO: 'enterpriseSearch', 'workplaceSearch'
+      catalogue: ['appSearch'], // TODO: 'enterpriseSearch', 'workplaceSearch'
+      privileges: null,
+    });
+
+    /**
+     * Register user access to the Enterprise Search plugins
+     */
+    capabilities.registerSwitcher(
+      async (request: KibanaRequest, uiCapabilities: UICapabilities) => {
+        const dependencies = { config, security, request, log: this.logger };
+
+        const { hasAppSearchAccess } = await checkAccess(dependencies);
+        // TODO: hasWorkplaceSearchAccess
+
+        return {
+          ...uiCapabilities,
+          navLinks: {
+            ...uiCapabilities.navLinks,
+            appSearch: hasAppSearchAccess,
+          },
+          catalogue: {
+            ...uiCapabilities.catalogue,
+            appSearch: hasAppSearchAccess,
+          },
+        };
+      }
+    );
+
+    /**
+     * Register routes
+     */
+    const router = http.createRouter();
+    const dependencies = { router, config, log: this.logger };
+
+    registerPublicUrlRoute(dependencies);
+    registerEnginesRoute(dependencies);
+
+    /**
+     * Bootstrap the routes, saved objects, and collector for telemetry
+     */
+    savedObjects.registerType(appSearchTelemetryType);
+
+    getStartServices().then(([coreStart]) => {
+      const savedObjectsStarted = coreStart.savedObjects as SavedObjectsServiceStart;
+
+      registerTelemetryRoute({
+        ...dependencies,
+        getSavedObjectsService: () => savedObjectsStarted,
+      });
```

Comment:
This is effectively registering the telemetry route after the plugin's `start`. This is currently working because we are waiting for the legacy plugins to initialize before starting the server (which are initialized after KP plugin's `start`), however this may break after legacy removal if we decide to change this workflow.

I see that you can't just rely on the route handler context because you are using the contract to create a repository to use `incrementCounter`, however as a workaround, you could do:

```ts
let soStart: SavedObjectsServiceStart;
getStartServices().then(([coreStart]) => {
  soStart = coreStart.savedObjects
});
registerTelemetryRoute({
  ...dependencies,
  getSavedObjectsService: () => soStart,
});
```

@joshdover do we have any plan to allow access to `incrementCounter` from the SO client or from the SO context handler by any other mean?

Reply:
Thanks so much for the great catches & cleanup Pierre! 😁 This comment should be addressed in 9c6acdd

Reply:
> @joshdover do we have any plan to allow access to incrementCounter from the SO client or from the SO context handler by any other mean?

No plans that I'm aware of.

I'm not sure why this was originally omitted from the Client interface (this all predates me at Elastic), but I suspect there may be security / authz concerns due to the scripted update that `incrementCounter` uses. In this specific case, authz doesn't really matter since we're accessing a telemetry object that (I assume?) is space agnostic and only accessible to the kibana_system user. 

I think we don't have any internal repositories available on the RequestHandlerContext in order to prevent accidental usage when a scoped client is probably what you want. Another option would be to register your own context provider to supply this on the `context` object:

```
coreSetup.http.registerRouteHandlerContext('enterpriseSearch', () => {
  const [coreStart] = await coreSetup.getStartServices();
  return {
    internalSavedObjectsRepository: coreStart.savedObjects.createInternalRepository()
  }
});
```

This would make the value accessible on `context.enterpriseSearch.internalSavedObjectsRepository` inside your route handlers.